### ISSUE TITLE
[Proposal] Named Axes/Dimensions or Tensor Shape Annotations

### ISSUE THREAD

**Issue by michalwols** [pytorch/pytorch#4164]
It would be great if there was a way to name tensor dimensions and use the names in all of the tensor methods that take dim as an argument. 

```python
images = torch.Tensor(data, dims=('batch', 'channel', 'height', 'width'))
more_images = torch.Tensor(data, dims=('batch', 'channel', 'height', 'width'))

torch.cat((images, more_images), dim='batch')
```

This could obviously be expanded to support axis objects that could support a basic algebra to allow using dimensions in functions like `view()`.

Another somewhat related feature would be a way to add type annotations as proposed in [this google doc](https://docs.google.com/document/d/1vpMse4c6DrWH5rq2tQSx3qwP_m_0lyn-Ij4WHqQqRHY/edit#heading=h.rkj7d39awayl):

```python
Batch = DimensionVar('Batch')
Row = DimensionVar('Row')
Column = DimensionVar('Column')
Channel = DimensionVar('Channel', size=3)

RGBImage = NDArray[Batch, Row, Column, Channel]
GrayscaleImage = NDArray[Batch, Row, Column]

def rgb_to_grayscale(array: RGBImage) -> GrayscaleImage:
  return numpy.mean(array, axis=RGBImage.axis(Channel))
```

- [Datarray: numpy with axis names](https://github.com/BIDS/datarray)
- [python mailing list discussion](https://mail.python.org/pipermail/numpy-discussion/2017-November/077429.html)
- [Google doc proposing array shape annotations](https://docs.google.com/document/d/1vpMse4c6DrWH5rq2tQSx3qwP_m_0lyn-Ij4WHqQqRHY/edit#heading=h.rkj7d39awayl)
- [Python PEP 472 proposing named indexing]( https://www.python.org/dev/peps/pep-0472/)

**Comment by apaszke** [comment]
That's an interesting idea that we might want to explore at some point. I'll leave it open

**Comment by ozancaglayan** [comment]
I came to open an issue and found this =)
Even without modifying the API to add `dims=()` as in the request of the Op, maybe it's possible to monkey patch torch in runtime so that  `torch.Size().__repr__()` prints `torch.Size([batch=B, seq=S, dim=D])` if there's an attribute say `_dim_names` in the Tensor?

**Comment by c0nn3r** [comment]
Implementing lots of transformer models has further convinced me this is a really good idea. Often you see people writing out the dimensions names in a comment on top of the line they are working on.

However, the only way I can see this idea being competitive in speed is manipulating the AST (while respecting transposes).

**Comment by obust** [comment]
The following articles gather a lot of the arguments backing up the named dimensions proposal

* [Tensor Considered Harmful Part 1](http://nlp.seas.harvard.edu/NamedTensor)
* [Tensor Considered Harmful Part 2](http://nlp.seas.harvard.edu/NamedTensor2)

**Comment by soumith** [comment]
@obust those are great blogposts that made us move on this. @zou3519 is actively working on it now.

**Comment by makslevental** [comment]
@zou3519 is there any progress on this?

**Comment by zou3519** [comment]
> @zou3519 is there any progress on this?

We released named tensors as an experimental feature for pytorch 1.3: https://pytorch.org/docs/stable/named_tensor.html. It's now possible to name tensor dimensions and use the names in tensor methods in a growing number of functions.

**Comment by NumesSanguis** [comment]
Just trying to connect 2 libraries here.

How does PyTorch's named dimensions implementation differ from xarray (http://xarray.pydata.org/en/latest/why-xarray.html)?
> Xarray introduces labels in the form of dimensions, coordinates and attributes on top of raw NumPy-like multidimensional arrays, which allows for a more intuitive, more concise, and less error-prone developer experience.

Over there, they're also talking about integrating with PyTorch, to leverage GPU power:
https://github.com/pydata/xarray/issues/3232

Maybe some conversation between you 2 can have beneficial results for the Python ecosystem ^-^

**Comment by ngimel** [comment]
Closing per @zou3519 comment.
